src/textual/widgets/_data_table.py

```diff
@@ -1767,58 +1845,113 @@ def _render_cell(
-                if is_header_cell or is_row_label_cell:
-                    # Apply subtle variation in style for the header/label (blue background by
-                    # default) rows and columns affected by the cursor, to ensure we can
-                    # still differentiate between the labels and the data.
-                    component_style += get_component("datatable--header-hover")
-
-            if cursor and show_cursor:
-                cursor_style = get_component("datatable--cursor")
-                component_style += cursor_style
-                if is_header_cell or is_row_label_cell:
-                    component_style += get_component("datatable--header-cursor")
-                elif is_fixed_style_cell:
-                    component_style += get_component("datatable--fixed-cursor")
-
-            post_foreground = (
-                Style.from_color(color=component_style.color)
-                if self.cursor_foreground_priority == "css"
-                else Style.null()
-            )
-            post_background = (
-                Style.from_color(bgcolor=component_style.bgcolor)
-                if self.cursor_background_priority == "css"
-                else Style.null()
+            component_style, post_style = self._get_styles_to_render_cell(
+                is_header_cell,
+                is_row_label_cell,
+                is_fixed_style_cell,
+                hover,
+                cursor,
+                self.show_cursor,
+                self._show_hover_cursor,
+                self.cursor_foreground_priority == "css",
+                self.cursor_background_priority == "css",
             )
 
+            if is_header_cell:
+                options = self.app.console.options.update_dimensions(
+                    width, self.header_height
+                )
+            else:
+                row = self.rows[row_key]
+                # If an auto-height row hasn't had its height calculated, we don't fix
+                # the value for `height` so that we can measure the height of the cell.
+                if row.auto_height and row.height == 0:
+                    options = self.app.console.options.update_width(width)
+                else:
+                    options = self.app.console.options.update_dimensions(
+                        width, row.height
+                    )
             lines = self.app.console.render_lines(
                 Styled(
                     Padding(cell, (0, 1)),
                     pre_style=base_style + component_style,
-                    post_style=post_foreground + post_background,
+                    post_style=post_style,
                 ),
-                self.app.console.options.update_dimensions(width, height),
+                options,
             )
 
             self._cell_render_cache[cell_cache_key] = lines
 
         return self._cell_render_cache[cell_cache_key]
 
+    @functools.lru_cache(
+        maxsize=512
```

Comment:
This is broken I'm afraid. @lru_cache on a method will give you a *single* cache for the process. You are using `get_component` which may return different results depending on the instance. Consequently if you have 2 or more DataTables, the styles are going to get mixed.

If you need caching (suspect it is a win), have a look at `cache.py` which has some home grown cache objects you can use per-instance.

Also consider reducing the maxsize. I'm guessing some combos are used more than others.

Reply:
While we concluded that this is not broken in the sense that you described because `lru_cache` also caches the instance, there are issues waiting to happen with hot-reloading, dark-mode toggling, etc.
Those are mostly addressed in #3313 and 2524af4.

I'd like for #3313 to be merged into main so I can merge it here to make sure this was all I needed and to add tests for it.